parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,412 @@
+   * used by bloom filter.
+   */
+  public final double DEFAULT_FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  public static final int HEADER_SIZE = 12;
+
+  // Bytes in a bucket.
+  public static final int BYTES_PER_BUCKET = 32;
+
+  // Hash strategy used in this bloom filter.
+  public final HASH hash;
+
+  // Algorithm applied of this bloom filter.
+  public final ALGORITHM algorithm;
+
+  // The underlying byte array for bloom filter bitset.
+  private byte[] bitset;
+
+  // A integer array buffer of underlying bitset help setting bits.
+  private IntBuffer intBuffer;
+
+  // A cache to column distinct value (hash)
+  private Set<Long> elements;
+
+  /**
+   * Constructor of bloom filter, if numBytes is zero, bloom filter bitset
+   * will be created lazily and the number of bytes will be calculated through
+   * distinct values in cache.
+   * @param numBytes The number of bytes for bloom filter bitset, set to zero can
+   *                 let it calculate number automatically by using default DEFAULT_FPP.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(int numBytes, HASH hash, ALGORITHM algorithm) {
+    if (numBytes != 0) {
+      initBitset(numBytes);
+    } else {
+      this.elements = new HashSet<>();
+    }
+    this.hash = hash;
+    this.algorithm = algorithm;
+  }
+
+  /**
+   * Construct the bloom filter with given bit set, it is used
+   * when reconstruct bloom filter from parquet file.
+   * @param bitset The given bitset to construct bloom filter.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(byte[] bitset, HASH hash, ALGORITHM algorithm) {
+    this.bitset = bitset;
+    this.intBuffer = ByteBuffer.wrap(bitset).asIntBuffer();
+    this.hash = hash;
+    this.algorithm = algorithm;
+  }
+
+  /**
+   * Construct the bloom filter with given bit set.
+   * @param input The given bitset to construct bloom filter.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(ByteBuffer input, HASH hash, ALGORITHM algorithm) {
+    if (input.isDirect()){
+      bitset = new byte[input.remaining()];
+      input.get(bitset);
+    } else {
+      this.bitset = input.array();
+    }
+    this.intBuffer = ByteBuffer.wrap(bitset).asIntBuffer();
+    this.hash = hash;
+    this.algorithm = algorithm;
+  }
+
+  /**
+   * Create a new bitset for bloom filter, at least 256 bits will be create.
+   * @param numBytes number of bytes for bit set.
+   */
+  private void initBitset(int numBytes) {
+    if (numBytes < BYTES_PER_BUCKET) {
+      numBytes = BYTES_PER_BUCKET;
+    }
+
+    // Get next power of 2 if it is not power of 2.
+    if ((numBytes & (numBytes - 1)) != 0) {
+      numBytes = Integer.highestOneBit(numBytes) << 1;
```

Comment:
This could be negative if a bit gets shifted into the highest-order position, yes?

Reply:
yes, I updated overflow handle code following.